src/core/shaderlib/lighting/lighting.js

```diff
@@ -20,9 +20,131 @@
 
 import lightingShader from './lighting.glsl';
 import project from '../project/project';
+import {COORDINATE_SYSTEM} from '../../lib/constants';
+import vec4_transformMat4 from 'gl-vec4/transformMat4';
+import vec3_sub from 'gl-vec3/subtract';
 
 export default {
   name: 'lighting',
   dependencies: [project],
-  vs: lightingShader
+  vs: lightingShader,
+  getUniforms
 };
+
+const INITIAL_MODULE_OPTIONS = {};
+
+const DEFAULT_LIGHTS_POSITION = [-122.45, 37.75, 8000];
+const DEFAULT_LIGHTS_STRENGTH = [2.0, 0.0];
+const DEFAULT_AMBIENT_RATIO = 0.05;
+const DEFAULT_DIFFUSE_RATIO = 0.6;
+const DEFAULT_SPECULAR_RATIO = 0.8;
+
+function getUniforms(opts = INITIAL_MODULE_OPTIONS) {
+  if (!opts.lightSettings) {
+    return {};
+  }
+
+  const {
+    numberOfLights = 1,
+
+    lightsPosition = DEFAULT_LIGHTS_POSITION,
+    lightsStrength = DEFAULT_LIGHTS_STRENGTH,
+    coordinateSystem = COORDINATE_SYSTEM.LNGLAT,
+    coordinateOrigin = [0, 0, 0],
+    modelMatrix = null,
+
+    ambientRatio = DEFAULT_AMBIENT_RATIO,
+    diffuseRatio = DEFAULT_DIFFUSE_RATIO,
+    specularRatio = DEFAULT_SPECULAR_RATIO
+  } = opts.lightSettings;
+
+  const projectionParameters = {
+    viewport: opts.viewport,
+    coordinateSystem,
+    coordinateOrigin,
+    layerCoordinateSystem: opts.coordinateSystem,
+    layerCoordinateOrigin: opts.coordinateOrigin,
+    modelMatrix
+  };
+
+  const lightsPositionWorld = [];
+  for (let i = 0; i < numberOfLights; i++) {
+    const position = preProject(lightsPosition.slice(i * 3, i * 3 + 3), projectionParameters);
+
+    lightsPositionWorld[i * 3] = position[0];
+    lightsPositionWorld[i * 3 + 1] = position[1];
+    lightsPositionWorld[i * 3 + 2] = position[2];
+  }
+
+  return {
+    lighting_lightPositions: lightsPositionWorld,
+    lighting_lightStrengths: lightsStrength,
+    lighting_ambientRatio: ambientRatio,
+    lighting_diffuseRatio: diffuseRatio,
+    lighting_specularRatio: specularRatio,
+    lighting_numberOfLights: numberOfLights
+  };
+}
+
+/* Projection utils */
+
+function lngLatZToWorldPosition(lngLatZ, viewport) {
+  const [X, Y] = viewport.projectFlat(lngLatZ);
+  const Z = (lngLatZ[2] || 0) * viewport.scale;
+  return [X, Y, Z];
+}
+
+function preProject(
+  position,
+  {
+    viewport,
+    coordinateSystem,
+    coordinateOrigin,
+    modelMatrix,
+    layerCoordinateSystem,
+    layerCoordinateOrigin
+  }
+) {
+  let [x, y, z] = position;
+  let worldPosition;
+
+  if (modelMatrix) {
+    [x, y, z] = vec4_transformMat4([], [x, y, z, 1.0], modelMatrix);
+  }
+
+  // pre-project light coordinates
+  switch (coordinateSystem) {
+    case COORDINATE_SYSTEM.LNGLAT:
```

Comment:
Ok, I hope I can state this clearly:

The use of the `COORDINATE_SYSTEM` constant in the lighting module is a reverse dependency on the `core/src/lib` "sub-module". Suggest we move the constant to `src/viewports` and just import it into `lib` etc.

In other words, to maintain encapsulation and avoid a situation where everything imports from and knows about everything, I think we should carve out a section of the code (`COORDINATE_SYSTEM` constant, `Viewport` hierarchy, `project*` and `lighting` shader modules) that know about each other, but they should not know about `src/lib` or any other code.

Also, since lighting has knowledge of the details of how to project that makes this module at least conceptually dependent on the viewport system and project module, so I think we could use shader functions or JS helper functions from project here - maybe that could simplify some code.